cli/args/mod.rs

```diff
@@ -353,6 +353,118 @@ fn resolve_lint_rules_options(
+    >,
+    node_resolver: Arc<crate::node::CliNodeResolver>,
+  ) -> Self {
+    Self {
+      workspace_resolver,
+      node_resolver,
+    }
+  }
+}
+impl MainModuleResolver for WorkspaceMainModuleResolver {
+  fn resolve_main_module(
+    &self,
+    specifier: &str,
+    cwd: &Url,
+  ) -> Result<Url, AnyError> {
+    let resolution = self.workspace_resolver.resolve(
+      &specifier,
+      &cwd,
+      deno_resolver::workspace::ResolutionKind::Execution,
+    )?;
+    let url = match resolution {
+      deno_resolver::workspace::MappedResolution::Normal {
+        specifier, ..
+      } => specifier,
+      deno_resolver::workspace::MappedResolution::WorkspaceJsrPackage {
+        specifier,
+        ..
+      } => specifier,
+      deno_resolver::workspace::MappedResolution::WorkspaceNpmPackage {
+        target_pkg_json,
+        sub_path,
+        ..
+      } => self
+        .node_resolver
+        .resolve_package_subpath_from_deno_module(
+          target_pkg_json.clone().dir_path(),
+          sub_path.as_deref(),
+          Some(cwd),
+          node_resolver::ResolutionMode::Import,
+          node_resolver::NodeResolutionKind::Execution,
+        )?
+        .into_url()?,
+      deno_resolver::workspace::MappedResolution::PackageJson {
+        sub_path,
+        dep_result,
+        alias,
+        ..
+      } => {
+        let result = dep_result
+          .as_ref()
+          .map_err(|e| deno_core::anyhow::anyhow!("{e}"))?;
+        match result {
+          deno_package_json::PackageJsonDepValue::File(file) => {
+            let cwd_path = deno_path_util::url_to_file_path(cwd)?;
+            deno_path_util::resolve_path(file, &cwd_path)?
+          }
+          deno_package_json::PackageJsonDepValue::Req(package_req) => {
+            ModuleSpecifier::parse(&format!(
+              "npm:{}{}",
+              package_req,
+              sub_path.map(|s| format!("/{}", s)).unwrap_or_default()
+            ))?
+          }
+          deno_package_json::PackageJsonDepValue::Workspace(version_req) => {
+            let pkg_folder = self
+              .workspace_resolver
+              .resolve_workspace_pkg_json_folder_for_pkg_json_dep(
+                alias,
+                version_req,
+              )?;
+            self
+              .node_resolver
+              .resolve_package_subpath_from_deno_module(
+                pkg_folder,
+                sub_path.as_deref(),
+                Some(cwd),
+                node_resolver::ResolutionMode::Import,
+                node_resolver::NodeResolutionKind::Execution,
+              )?
+              .into_url()?
+          }
+          deno_package_json::PackageJsonDepValue::JsrReq(_) => {
+            return Err(
+              deno_resolver::DenoResolveErrorKind::UnsupportedPackageJsonJsrReq
+                .into_box()
+                .into(),
+            )
+          }
+        }
+      }
+    };
+    Ok(url)
+  }
+}
+
```

Comment:
This code should probably be moved to somewhere else, I just don't have a strong opinion on where